components/scream/src/physics/p3/p3_functions_dsd2_impl.hpp

```diff
@@ -0,0 +1,130 @@
+      mu_c_local = pack::max(mu_c_local, 2);
+      mu_c_local = pack::min(mu_c_local, 15);
+
+      mu_c.set(qc_gt_small, mu_c_local);
+    }
+
+    // interpolate for mass distribution spectral shape parameter (for SB warm processes)
+    if (P3C::iparam == 1) {
+      IntSmallPack dumi = IntSmallPack(mu_c) - 1;
+      Spack dnu0, dnu1;
+      pack::index_and_shift<1>(dnu, dumi, dnu0, dnu1);
+      nu.set(qc_gt_small, dnu0 + (dnu1 - dnu0) * (mu_c - Spack(dumi) - 1));
+    }
+
+    // calculate lamc
+    lamc.set(qc_gt_small, pack::cbrt(cons1 * nc * (mu_c + 3) * (mu_c + 2) * (mu_c + 1) / qc));
+
+    // apply lambda limiters
+    Spack lammin = (mu_c + 1)*sp(2.5e+4); // min: 40 micron mean diameter
+    Spack lammax = (mu_c + 1)*sp(1.e+6);   // max:  1 micron mean diameter
+
+    Smask lamc_lt_min = lamc < lammin && qc_gt_small;
+    Smask lamc_gt_max = lamc > lammax && qc_gt_small;
+    Smask min_or_max = lamc_lt_min || lamc_gt_max;
+    lamc.set(lamc_lt_min, lammin);
+    lamc.set(lamc_gt_max, lammax);
+
+    nc.set(min_or_max, 6 * (lamc * lamc * lamc) * qc / (C::Pi * C::RHOW * (mu_c + 3) * (mu_c + 2) * (mu_c + 1)));
+
+    cdist.set(qc_gt_small, nc * (mu_c+1) / lamc);
+    cdist1.set(qc_gt_small, nc * lcldm / pack::tgamma(mu_c + 1));
+  }
+}
+
+template <typename S, typename D>
+void Functions<S,D>::
+get_rain_dsd2 (
+    const Smask& qr_gt_small, const Spack& qr, Spack& nr, Spack& mu_r,
+    Spack& lamr, Spack& cdistr, Spack& logn0r, const Spack& rcldm)
+{
+  constexpr auto nsmall = Constants<Scalar>::NSMALL;
+  constexpr auto cons1 = Constants<Scalar>::CONS1;
+
+  lamr = 0;
+  cdistr = 0;
+  logn0r = 0;
+
+  if (qr_gt_small.any()) {
+    constexpr Scalar mu_r_const = C::mu_r_const;
+    // use lookup table to get mu
+    // mu-lambda relationship is from Cao et al. (2008), eq. (7)
+
+    // find spot in lookup table
+    // (scaled N/q for lookup table parameter space)
+    const auto nr_lim = max(nr, nsmall);
+    Spack inv_dum(0);
+    inv_dum.set(qr_gt_small, pack::cbrt(qr / (cons1 * nr_lim * 6)));
+
+    // Apply constant mu_r:  Recall the switch to v4 tables means constant mu_r
+    mu_r.set(qr_gt_small, mu_r_const);
+    // recalculate slope based on mu_r
+    lamr.set(qr_gt_small, pack::cbrt(cons1 * nr_lim * (mu_r + 3) *
+                                     (mu_r + 2) * (mu_r + 1)/qr));
+
+    // check for slope
+    const auto lammax = (mu_r+1.)*sp(1.e+5);
+    // set to small value since breakup is explicitly included (mean size 0.8 mm)
+    const auto lammin = (mu_r+1.)*sp(1250.0);
+
+    // apply lambda limiters for rain
+    const auto lt = qr_gt_small && (lamr < lammin);
+    const auto gt = qr_gt_small && (lamr > lammax);
+    const auto either = lt || gt;
+    nr.set(qr_gt_small, nr_lim);
+    if (either.any()) {
+      lamr.set(lt, lammin);
+      lamr.set(gt, lammax);
+      scream_masked_loop(either, s) {
+        nr[s] = std::exp(3*std::log(lamr[s]) + std::log(qr[s]) +
```

Comment:
Now that I think about it, I used to think all the std functions were not cuda friendly, and we should use, say, `exp` instead of `std::exp`. Why isn't this causing warnings like "calling a __host__ function from a __device__ function is not allowed" on cuda?

Reply:
I'm almost certain CUDA is smart enough to inject it's own versions such that they get used in kernels.